tutorials/3_transactions.md

```diff
@@ -99,127 +148,139 @@ struct Fund {
+inside the contract which calls `new`. To illustrate, if we defined a contract
+`struct Manager contract_0`, then the balance of `fund` is
+`contract_0.user_fund.model_address`.
 
-Now navigate to `void run_model(void)`. This method is the entry-point to our
-model. After some bootstrapping, we can see `Manager` initialized:
+Between lines 21 to 86, we then find the encoding of each method. To avoid name
+collisions, SmartACE renames each method according to the following rules:
 
-```cpp
-smartace_log("[Initializing contract_0]");
-((sender).v) = (rt_nd_range(3, 5, "sender"));
-((value).v) = (0);
-((blocknum_tmp).v) = (nd_uint256_t("blocknum"));
-sol_require(((blocknum).v) <= ((blocknum_tmp).v), 0);
-(blocknum) = (blocknum_tmp);
-((timestamp_tmp).v) = (nd_uint256_t("timestamp"));
-sol_require(((timestamp).v) <= ((timestamp_tmp).v), 0);
-(timestamp) = (timestamp_tmp);
-Init_Manager(&(contract_0), sender, value, blocknum, timestamp, paid, origin);
-```
+  * The constructor of contract `Manager` maps to `Manager_Constructor`.
+  * The fallback method of `Manager` maps to `Manager_Fallback`.
+  * The method `openFund` of `Manager` maps to `Manager_Method_openFund`.
 
-(*SW: `rt_` is an old acronym that made sense at the time. I think we should find something new. Perhaps just `nd_range` and `nd_byte`*).
+We then move on to the entry-point, `void run_model(void)`, at line 85. After
+some setup code, we find at line 102 the initialization of `Manager`, and in
+turn, `Manager.fund`.
 
-Each method prefixed by `nd_` is used to generate a non-deterministic value. In
-this code we construct `Manager` using a non-deterministic `msg.sender`,
-`msg.value`, `block.number`, and `block.timestamp`. Therefore, the balance of
-`fund` is `contract_0.user_fund.user_balances`.
+```cpp
+smartace_log("[Initializing contract_0 and children]");
+sender.v = nd_range(3, 5, "sender");
+value.v = 0;
+blocknum.v = nd_increase(blocknum.v, 0, "blocknum");
+timestamp.v = nd_increase(timestamp.v, 0, "timestamp");
+Init_Manager(&contract_0, sender, value, blocknum, timestamp, paid, origin);
+```
 
-This sets up an arbitrary instance of `Manager` using non-deterministic inputs.
-In particular, we model `msg.sender`, `msg.value`, an increasing `block.number`,
-and an increasing `block.timestamp`. The astute may notice that we only select
-from addresses 3 and 4, inclusively. We will show in a later post why this is
-sufficient for soundness.
+To summarize briefly, each method prefixed by `nd_` selects a value
+non-deterministically. These values are used to model an arbitrary `msg.sender`,
+and an increasing `block.number` and `block.timestamp`. An attentive reader may
+notice that we restrict addresses to 3 and 4. The importance of this choice, and
+its impact on completeness, are the topic of the next tutorial. For now we can
+think of this as a bounded model with two clients, so at the least, our
+verification is sound.
```

Comment:
When I read this it seems surprising that the verification would be sound if you have a bounded model.

Reply:
Good point. I might have soundness and completeness flipped here.

I was thinking wrt "the proof that a violation exists". In this context, soundness is "all detected violations are true violations" and completeness is "all true violations are detected."

I can flip this.

Reply:
👍 Yeah, these terms are sometimes used differently in different communities. However, in verification it's usually the other way around.